interface/src/app/analyze/analysis/sampleBin.js

```diff
@@ -301,36 +389,50 @@ function($log, $cacheFactory, $q, Sample, Activity, NodeInfo, MathFuncts) {
-          //      `activityCache` [inner .map()]
-          // FIXME: counting on array order to match node order here
-          return this.activityCache.get(sampId)[val.node - 1].value;
-        };
-        var nodeObject = {
-          'id': val.node, // TODO: map this id to node name (via vega spec?)
-          'activityA': sg['group-a'].map(mapSampleIdsToActivity, this),
-          'activityB': sg['group-b'].map(mapSampleIdsToActivity, this)
-        };
-        nodeObject.diff = (
-          MathFuncts.mean(nodeObject.activityB) -
-          MathFuncts.mean(nodeObject.activityA)
-        );
-        nodeObject.logsig = -Math.log10(MathFuncts.tTest(
-          nodeObject.activityA, nodeObject.activityB
-        ).pValue());
+      // (1a) we obtain a list of nodes by retrieving node activity
+      //      for the first sample in our volcano plot
+      var firstSampleNodes = this.activityCache.get(sg['group-a'][0]).map(
+        function(val) {
+          return val.node;  // extract just the node IDs
+        }
+      );
+      // (1b) now obtain (and cache) a name for each node id
+      var nodeInfoSetPromise = this.getNodeInfoSetPromise(firstSampleNodes);
+      var mapNodesToNodeInfo = function() {
+        // (2a) next, we build an array (replacing `volcanoData.source`)
+        //      comprised of `nodeObject`s by walking through the
+        //      `firstSampleNodes` and constructing a `nodeObject` for
+        //      each. [outer .map()]
+        cbSampleBin.volcanoData.source = firstSampleNodes.map(function(id) {
+          var mapSampleIdsToActivity = function(sampId) {
+            // (2b) the array of activity for each node is built by plucking the
+            //      activity `.value` for each sample within this node from the
+            //      `activityCache` [inner .map()]
+            // FIXME: counting on array order to match node order here
+            return cbSampleBin.activityCache.get(sampId)[id - 1].value;
+          };
+          var nodeObject = {
+            'id': id,
+            'name': cbSampleBin.getCachedNodeInfo(id).name,
+            'activityA': sg['group-a'].map(mapSampleIdsToActivity),
+            'activityB': sg['group-b'].map(mapSampleIdsToActivity)
+          };
+          nodeObject.diff = (
+            MathFuncts.mean(nodeObject.activityA) -
+            MathFuncts.mean(nodeObject.activityB)
+          );
+          nodeObject.logsig = -Math.log10(MathFuncts.tTest(
+            nodeObject.activityA, nodeObject.activityB
+          ).pValue());
 
-        return nodeObject;
-      }, this);
-      return retval;
+          return nodeObject;
+        });
+        // no return needed here: we've updated `cbSampleBin.volcanoData`
+      };
+      // invoke mapNodesToNodeInfo only after nodeInfoSetPromise is fulfilled
+      nodeInfoSetPromise
+        .then(mapNodesToNodeInfo)
+        .catch(this.logError);
     }
   };
```

Comment:
Also I would be in favor of leaving lines 433-435 as they are because it's more readable and idiomatic. Besides the fact that this function will barely fit onto a single screen I'm not actually worried about the line count of the function, my concern is the amount of functionality crammed into a single function.